ramp4-pcar4/storylines-editor
# re-implement rename feature

### Related Item(s)
#253 
#418 
#419

### Changes
- Added the `/rename` server endpoint: this is a re-implementation of the rename feature. Instead of relying on the `/upload` endpoint that simply created a new product folder, the `/rename` endpoint calls the `rename` syscall on the existing product folder. It will also delete and re-instantiate the Git repo.
- Added a new `/check` endpoint to the server to see if a product already exists. This way, we don't need to call the `/retrieve` endpoint and wait for the entire product to transfer every time the user enters a UUID. It should be much faster now.
- Moved the `responseMessages.push()` calls in to the `logger` function to reduce code duplication.

### Notes
![image](https://github.com/user-attachments/assets/6870b2a4-be1b-4dc7-ba46-016b74c2604f)

### Testing
This is a big one, so please test thoroughly! If my instructions below are confusing please reach out, or post a comment on this PR to help others if they also have trouble.

**:warning: IMPORTANT! :warning:** If testing locally, please move the `server` folder outside of the `storylines-editor` folder. `npm run dev` seems to lock the `server/public` folder and it will result in the `rename` system call hanging indefinitely otherwise. Alternatively, if you overwrite the `server/index.js` code with the new code in this PR, you should be able to test using the demo page associated with this PR.<br />
**:warning:** If you notice the rename endpoint is hanging at any point, then the product folder is being accessed from somewhere else at the same time. Try closing anything that could be using it (VS Code, File Explorer, notepad, etc.). As far as I understand this should (hopefully) not happen on prod as nothing should be accessing the file system. If anyone knows of something that can be added to the code to fix this, please let me know.


Steps:
1. Pull this PR locally and follow the important step above.
2. Try creating a new product with an existing UUID. Notice that the "UUID is already in use" error should appear much faster than before.
3. Create a new product that doesn't already exist. Save it so it gets uploaded to the server.
4. Return to the home page, and now load this new product.
5. Try to rename it, and ensure it works properly.
6. Now, try to load an existing product.
7. Try to rename it and ensure that everything works here as well. Also, check that the previous editing history has been wiped.
8. Try whatever you want to break this otherwise!


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ramp4-pcar4/storylines-editor/487)
